8350209: Preserve adapters in AOT cache

vnkozlov opened the pull request:
[JEP 483](https://bugs.openjdk.org/browse/JDK-8315737) preserves class information in AOT cache which helps Java startup performance.

We should also preserve adapters (i2c, c2i) to further improve performance of class linking where adapters are generated.

Short running Java application can see several percents improvement. I got 6% improvement when ran `HelloWorld.java` on Linux-x64 Ice Lake CPU (2.5Ghz):

```
(perf stat -r 100 java -XX:AOTCache=app.aotcache -cp hello.jar HelloWorld > /dev/null) 2>&1 | grep elapsed
         0.0299401 +- 0.0000504 seconds time elapsed  ( +-  0.17% )

(perf stat -r 100 java -XX:AOTCache=app.aotcache -XX:+UnlockDiagnosticVMOptions -XX:-AOTAdapterCaching -cp hello.jar HelloWorld > /dev/null) 2>&1 | grep elapsed
         0.0318654 +- 0.0000535 seconds time elapsed  ( +-  0.17% )
```

New diagnostic flags are introduced (use `-XX:+UnlockDiagnosticVMOptions` to unlock them):

```
-XX:+AOTAdapterCaching  - Enable or disable saving and restoring i2c2i adapters
-XX:AOTCodeMaxSize=10*M - buffer size in bytes for AOT code caching
-XX:+AbortVMOnAOTCodeFailure - Abort VM on the first occurrence of AOT code caching failure
```
By default `AOTAdapterCaching` is `false` and enabled ergonomically when `-XX:AOTCache` is specified.
This flag is ignored when `AOTCache` is not specified.

To use AOT adapters follow process described in JEP 483:

```
java -XX:AOTMode=record -XX:AOTConfiguration=app.aotconf -cp app.jar App
java -XX:AOTMode=create -XX:AOTConfiguration=app.aotconf -XX:AOTCache=app.aot -cp app.jar
java -XX:AOTCache=app.aot -cp app.jar App
```

There are several new UL flag combinations to trace the AOT code caching process:

```
-Xlog:aot+codecache+init -Xlog:aot+codecache+exit -Xlog:aot+codecache+stubs
```

@ashu-mehra is main author of changes. He implemented adapters caching.
I did main framework (`AOTCodeCache` class) for saving and loading AOT code.

Tested tier1-6,10, which includes tests with `AOTClassLinking` enabled. Also Xcomp,stress and JCK.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8350209](https://bugs.openjdk.org/browse/JDK-8350209): Preserve adapters in AOT cache (**Enhancement** - P3)


### Reviewers
 * [Vladimir Ivanov](https://openjdk.org/census#vlivanov) (@iwanowww - **Reviewer**)
 * [Magnus Ihse Bursie](https://openjdk.org/census#ihse) (@magicus - **Reviewer**) Review applies to [e29da02e](https://git.openjdk.org/jdk/pull/24740/files/e29da02e19830511c7af57c17e3e0cdc1b4418a5)
 * [Ashutosh Mehra](https://openjdk.org/census#asmehra) (@ashu-mehra - Committer)
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**) Review applies to [78f2828d](https://git.openjdk.org/jdk/pull/24740/files/78f2828d34b4a7b0c864489c67f8c37a61c26f46)

### Contributors
 * Ashutosh Mehra `<asmehra@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/24740/head:pull/24740` \
`$ git checkout pull/24740`

Update a local copy of the PR: \
`$ git checkout pull/24740` \
`$ git pull https://git.openjdk.org/jdk.git pull/24740/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 24740`

View PR using the GUI difftool: \
`$ git pr show -t 24740`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/24740.diff">https://git.openjdk.org/jdk/pull/24740.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/24740#issuecomment-2813973131)
</details>

vnkozlov commented:
/contributor add @ashu-mehra

ashu-mehra commented:
@vnkozlov I forgot to remove `AOTCodeSection` from `code/aotCodeCache.hpp` when I refactored the APIs. This struct is no longer used. Can you please remove it.

vnkozlov commented:
> @vnkozlov I forgot to remove `AOTCodeSection` from `code/aotCodeCache.hpp` when I refactored the APIs. This struct is no longer used. Can you please remove it.

Done.

vnkozlov commented:
/cc hotspot-runtime
/cc hotspot-compiler

vnkozlov commented:
/label remove hotspot

iklam commented on src/hotspot/share/cds/cdsConfig.cpp:
```diff
@@ -847,3 +857,22 @@ bool CDSConfig::is_dumping_method_handles() {
+
+static bool _is_dumping_aot_code_enabled = false;
+
+bool CDSConfig::is_dumping_aot_code_enabled() {
+  return _is_dumping_aot_code_enabled;
```
Other functions in CDSConfig don't have the `_enabled()` suffix. E.g., `is_dumping_method_handles()`. It doesn't mean we are doing method handle right now, but rather we have the ability to do so.

Since we rarely ask "am I in the middle of dumping X right now", I think adding `_enabled()` will be redundant.

vnkozlov replied:
Okay.

vnkozlov commented:
Added missing changes for Aarch64 to always generate far jumps in AOT code. `generate_i2c2i_adapters()` use `far_jump()`.
Note, for x64 `ForceUnreachable` flag is set to `true` in `AOTCodeCache::initialize()`. Unfortunately Aarch64 does not use this flag, so I need to add check to each query functions.

erikj79 commented:
Build change looks trivially good.

vnkozlov commented:
Thank you, @erikj79

iwanowww commented on src/hotspot/share/code/aotCodeCache.cpp:
```diff
@@ -0,0 +1,1283 @@
+  if ((entry_kind == AOTCodeEntry::Adapter) && !cache->adapter_caching()) {
+    return false;
+  }
+  log_info(aot, codecache, stubs)("Writing blob '%s' to AOT Code Cache", name);
```
I'd revisit logging code in AOTCodeCache and downgrade info->debug and debug->trace where appropriate. It feels too low-level most of the time.

vnkozlov replied:
done

iwanowww commented on src/hotspot/share/runtime/sharedRuntime.cpp:
```diff
@@ -2612,110 +2741,290 @@ AdapterHandlerEntry* AdapterHandlerLibrary::get_adapter(const methodHandle& meth
+  if (AOTCodeCache::is_using_adapters()) {
+    adapter_blob = AdapterHandlerLibrary::link_aot_adapter_handler(this);
+    if (adapter_blob == nullptr) {
+      log_warning(cds)("Failed to link AdapterHandlerEntry (fp=%s) to its code in the AOT code cache", _fingerprint->as_basic_args_string());
```
Doesn't it add noise in the output for not yet seen adapter shapes? It's a warning.

ashu-mehra replied:
`AdapterHandlerEntry::link` method gets called only for the archived adapters. Not yet seen adapters do not come across this code path. Reason for making it a warning is the assumption that, if the AOT Code Cache is usable, then we should always be able to link the `AdapterHandlerEntry` to its code in the AOT code cache. If we fail to do so, then something is not right in this machinery and likely pointing to a bug.

iwanowww commented on src/hotspot/share/code/aotCodeCache.cpp:
```diff
@@ -0,0 +1,1283 @@
+  AOTAdapterCaching = false;
+}
+
+static void exit_vm_on_store_failure() {
```
It's a bit confusing to see `exit_vm_on_load_failure()` and `exit_vm_on_store_failure()` to silently proceed unless a flag is explicitly specified.

Moreover, how reliable `AOTAdapterCaching = false` to fail-fast and avoid repreated load/store attempts? At least, I see that `AOTCodeCache` ctor cache `AOTAdapterCaching`, so it won't see the update.

How does it affect adapter code generation during assembly phase?

vnkozlov replied:
We check failure state of AOT code cache when query about using adapters caching:
```
  bool for_use()  const { return _for_use  && !_failed; }
  bool for_dump() const { return _for_dump && !_failed; }
  static bool is_on() CDS_ONLY({ return _cache != nullptr && !_cache->closing(); }) NOT_CDS_RETURN_(false);
  static bool is_on_for_use()  { return is_on() && _cache->for_use(); }
  static bool is_on_for_dump() { return is_on() && _cache->for_dump(); }
```

```
  static bool is_dumping_adapters() { return is_on_for_dump() && _cache->adapter_caching(); }
  static bool is_using_adapters()   { return is_on_for_use() && _cache->adapter_caching(); }
```

vnkozlov replied:
AOT adapters code caching is loading is guarded by the methods not by flag.

Setting AOTAdapterCaching to false on failure is simple indication that adapter caching is switched of for someone who will look on final state of flag.

vnkozlov replied:
I added `log_info()` to `exit_vm_on_*_failure()` methods to produce notification when AbortVMOnAOTCodeFailure flag is off (default value).

iwanowww replied:
> Setting AOTAdapterCaching to false on failure is simple indication that adapter caching is switched off for someone who will look on final state of flag.

But how does it affect execution? What's the intended behavior when a failure happens during store attempt? What are the consistency guarantees for AOT code cache during dumping in presense of store failures? What I see right now is that errors reported by `AOTCodeCache::store_code_blob()` are silently ignored. How does `_cache` notice a store failure during dumping phase?

iwanowww reviewed (commented):
Finished the first pass over the code.


Overall, looks good. Some feedback follows.

iklam commented on src/hotspot/share/cds/aotCacheAccess.hpp:
```diff
@@ -0,0 +1,47 @@
+
+// AOT Cache API for AOT compiler
+
+class AOTCacheAccess : AllStatic {
```
This file is called https://github.com/openjdk/leyden/blob/premain/src/hotspot/share/cds/cdsAccess.hpp in the Leyden repo and provides an abstract API for accessing contents of the AOT cache. In Leyden, we have APIs for accessing cached oops:

```
  static int get_archived_object_permanent_index(oop obj) NOT_CDS_JAVA_HEAP_RETURN_(-1);
  static oop get_archived_object(int permanent_index) NOT_CDS_JAVA_HEAP_RETURN_(nullptr);
```

and various pointer operations

```
  static uint delta_from_shared_address_base(address addr);

  template <typename T>
  static void set_pointer(T** ptr, T* value) {
    set_pointer((address*)ptr, (address)value);
  }
  static void set_pointer(address* ptr, address value);
```

Let's keep the AOTCacheAccess name for now and wait until we merge this PR down to the Leyden repo. There's some overlap between AOTCacheAccess and CDSConfig. Maybe we should do a refactor/rename later.

vnkozlov replied:
Okay, I will keep the name.

vnkozlov commented on src/hotspot/share/cds/aotCacheAccess.hpp:
```diff
@@ -0,0 +1,47 @@
+
+class AOTCacheAccess : AllStatic {
+public:
+  static void* allocate_aot_code(size_t size) NOT_CDS_RETURN_(nullptr);
```
done

vnkozlov commented:
I found that C strings caching has issue - it is concurrent and needs synchronization.
Also I see a lot more strings after recent changes in mainline. After investigation I filed separate RFE for mainline and I am working on it: [JDK-8355635: Do not collect C strings in C2 scratch buffer](https://bugs.openjdk.org/browse/JDK-8355635)

vnkozlov commented:
Downgraded UL - minimum output for `log_info`.
Added new lock `AOTCodeCStrings_lock` to synchronize C strings caching - several compiler threads may add C strings simultaneously.
Also don't record c strings from C2 scratch buffer.

vnkozlov commented:
I repeated testing again - it passed.

@iwanowww, @ashu-mehra please review again.

vnkozlov commented:
Did new merge from mainline and fixed C strings caching issue found in `leyden/premain` branch.

ashu-mehra commented on src/hotspot/share/code/aotCodeCache.hpp:
```diff
@@ -330,6 +337,8 @@ class AOTCodeCache : public CHeapObj<mtCode> {
 
   static const char* add_C_string(const char* str) NOT_CDS_RETURN_(str);
 
+static const char* add_C_string2(const char* str) NOT_CDS_RETURN_(str);
```
This method is not used.

ashu-mehra commented:
Just to note there is still some code related to blobs but we are not storing them yet in the aot code cache. But I think we would need them pretty soon. So I think it should be okay to keep that code as is.

ashu-mehra commented:
rest lgtm

vnkozlov commented:
Thank you, @ashu-mehra, for review!

iklam commented on src/hotspot/share/cds/filemap.cpp:
```diff
@@ -1274,6 +1277,40 @@ char* FileMapInfo::map_bitmap_region() {
+    }
+    mapped_base = requested_base;
+  } else {
+    bool read_only = false, allow_exec = false;
```
Add comment for clarity:

```
// We do not execute in-place in the AOT code region. AOT code is copied to
// the CodeCache for execution.
bool read_only = false, allow_exec = false;
```

iklam commented on src/hotspot/share/include/cds.h:
```diff
@@ -35,7 +35,7 @@
 // Also, this is a C header file. Do not use C++ here.
 
-#define NUM_CDS_REGIONS 4 // this must be the same as MetaspaceShared::n_regions
+#define NUM_CDS_REGIONS 5 // this must be the same as MetaspaceShared::n_regions
```
Change `CURRENT_CDS_ARCHIVE_VERSION` below to `19`

iklam reviewed (approved):
LGTM. Some small nits.

vnkozlov commented:
Thank you, @iklam, for review. I pushed your suggestions.

vnkozlov commented on src/hotspot/share/runtime/sharedRuntime.cpp:
```diff
@@ -2612,110 +2741,290 @@ AdapterHandlerEntry* AdapterHandlerLibrary::get_adapter(const methodHandle& meth
+}
+
+#ifndef PRODUCT
+void AdapterHandlerLibrary::print_adapter_handler_info(AdapterHandlerEntry* handler, AdapterBlob* adapter_blob) {
```
Done.

vnkozlov commented on src/hotspot/share/code/relocInfo.hpp:
```diff
@@ -1287,6 +1289,24 @@ class trampoline_stub_Relocation : public Relocation {
 
   void pack_data_to(CodeSection * dest) override;
   void unpack_data() override;
+#if defined(AARCH64)
```
Good catch. I copied it from leyden/premain but it is used only for trampoline jump relocation for calls in nmethods on Aarch64. They are not present in adapters. I remove this code.

vnkozlov commented on src/hotspot/share/code/codeBlob.hpp:
```diff
@@ -188,6 +204,8 @@ class CodeBlob {
   nmethod* as_nmethod() const                 { assert(is_nmethod(), "must be nmethod"); return (nmethod*) this; }
   CodeBlob* as_codeblob() const               { return (CodeBlob*) this; }
+  AdapterBlob* as_adapter_blob() const        { assert(is_adapter_blob(), "must be adapter blob"); return (AdapterBlob*) this; }
+  ExceptionBlob* as_exception_blob() const    { assert(is_exception_stub(), "must be exception stub"); return (ExceptionBlob*) this; }
```
removed

vnkozlov commented on src/hotspot/share/code/aotCodeCache.cpp:
```diff
@@ -0,0 +1,1317 @@
+  if (AbortVMOnAOTCodeFailure) {
+    vm_exit_during_initialization("Unable to use AOT Code Cache.", nullptr);
+  }
+  log_info(aot, codecache, init)("Unable to use AOT Code Cache.");
```
changed

vnkozlov commented on src/hotspot/share/code/aotCodeCache.cpp:
```diff
@@ -0,0 +1,1317 @@
+    tty->print_cr("Unable to create AOT Code Cache.");
+    vm_abort(false);
+  }
+  log_info(aot, codecache, exit)("Unable to create AOT Code Cache.");
```
I will remove `exit_vm_on_` from these 2 methods to avoid confusion. And will add comment.

vnkozlov commented on src/hotspot/share/runtime/sharedRuntime.cpp:
```diff
@@ -2612,110 +2741,290 @@ AdapterHandlerEntry* AdapterHandlerLibrary::get_adapter(const methodHandle& meth
+    entry_offset[1] = handler->get_c2i_entry() - i2c_entry;
+    entry_offset[2] = handler->get_c2i_unverified_entry() - i2c_entry;
+    entry_offset[3] = handler->get_c2i_no_clinit_check_entry() - i2c_entry;
+    AOTCodeCache::store_code_blob(*adapter_blob, AOTCodeEntry::Adapter, id, name, AdapterHandlerEntry::ENTRIES_COUNT, entry_offset);
```
If something happened when we tried to store blob (adapter in this case) into buffer (for example, if reserved for AOT code buffer is too small) the next will be called:
```
    set_failed();
    exit_vm_on_store_failure();
```
So we either abort VM based on the flag or issue warning (as you suggested) and continue execution.

`set_failed()` will prevent following attempts to cache blobs and prevent final dump any cached code into AOT cache:
```
 bool for_dump() const { return _for_dump && !_failed; }
```

iwanowww replied:
Maybe add an assert here?

```
bool success = AOTCodeCache::store_code_blob(...);
assert(success || !AOTCodeCache::is_dumping_adapters(), "");
```

vnkozlov replied:
Thank you, @ashu-mehra. You have good points. I will work on them.

vnkozlov replied:
I will add assert message but keep check.

vnkozlov commented:
> How does _cache notice a store failure during dumping phase?
`set_failed()` sets flags AOTCodeCache::_failed. Which is checked when we trying to get _cache pointer:
```
```

vnkozlov commented:
Thank you, @iwanowww, for review. I think I addressed your comments and answered your question.

iwanowww commented on src/hotspot/share/code/aotCodeCache.cpp:
```diff
@@ -0,0 +1,1317 @@
+#include <sys/stat.h>
+#include <errno.h>
+
+static void load_failure() {
```
Maybe `on_(load|store)_failure()` or `report_(load|store)_failure()`?

vnkozlov replied:
I will use `report_`

iwanowww reviewed (approved):
Looks good!

Minor suggestions follow.

vnkozlov commented:
Thank you, @iwanowww. I need 2 re-reviews so I asked @ashu-mehra to review again.
Meanwhile I am re-running testing with latest mainline code.

vnkozlov commented:
I reverted back `log_warning()` to `log_info()` in `report_*_failure()` methods.

I removed `AOTCodeCache::_adapter_caching` field and use `AOTAdapterCaching` flag check instead. Setting the flag to `false` in case of a failure make more sense now.

iwanowww reviewed (approved):
Looks good.

vnkozlov commented:
/reviewers 1

ashu-mehra reviewed (approved):
lgtm

vnkozlov commented:
Thank you again, @iwanowww and @ashu-mehra

vnkozlov commented:
/integrate
